core/src/main/java/hudson/tasks/ArtifactArchiver.java

```diff
@@ -260,7 +260,7 @@ public void perform(Run<?,?> build, FilePath ws, EnvVars environment, Launcher l
                     f.setExcludes(excludes);
                     f.setDefaultExcludes(defaultExcludes);
                     f.setCaseSensitive(caseSensitive);
-                    f.perform(build, ws, launcher, listener);
+                    f.perform(build, ws, environment, launcher, listener);
```

Comment:
This might slightly change the behavior. Missing argument will force environment variable recalculation in https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/tasks/SimpleBuildStep.java#L91 . Instead of that, current code will use the environment passed in the beginning of the step. In the edge case `build.pickArtifactManager().archive` might perform some environment-contributing actions, which will no longer be considered after the change.

Reply:
@oleg-nenashev Yeah, I would agree. IIUC since `ArtifactManager` can be supplied by external storage plugins, it would be ideal to recalculate the environment.
```
f.perform(build, ws, run.getEnvironment(listener), launcher, listener);
```

Reply:
No, do not do that! The currently proposed patch is correct.